docs/design/uplink-contexts.md

```diff
@@ -0,0 +1,103 @@
+    EncryptionKey   []byte
+}
+```
+
+<blockquote>
+For concreteness, the following is an example UX around uplink with contexts. It is expected to evolve, and is just used to help with understanding.
+</blockquote>
+
+When setting up uplink for the first time, a default context is created, and the command will prompt you for some credentials. For example,
+
+```
+$ uplink setup
+A new "default" context has been created.
+
+Please respond with the type of information you have:
+1. API key from a satellite
+2. Exported context
+> 1
+
+API Key: <user pastes key>
+Satellite URL: <user puts in or picks satellite like now>
+```
+
+After a context has been created, bucket information must be added. For example,
+
+```
+$ uplink import [optionally specify a context with --context=default]
+
+Please respond with the type of information you have:
+1. Bucket with root passphrase
+2. Bucket share
+> 1
+
+Bucket name: <user inputs bucket name>
+Passphrase: <user inputs passphrase>
+
+$ uplink import
+Please respond with the type of information you have:restricted
+1. Bucket with root passphrase
+2. Bucket share
+> 2
+
+Bucket share: <user pastes shared bucket (RevealedPath)>
+```
+
+When a client attempts to interact with a bucket, it finds the latest added matching unencrypted path and appends the rest of the path, encrypted with the key the revealed path specifies, to the revealed path's encrypted path. For example, if one had this set of `BucketShare` objects:
+
+```go
+paths := []BucketShare{
+    {Bucket: "x", EncryptedPath: "/e1/e2/e3", UnecryptedPath: "/a/b/c", EncryptionKey: "key1"},
+    {Bucket: "x", EncryptedPath: "/e1",       UnecryptedPath: "/a",     EncryptionKey: "key2"},
+    {Bucket: "x", EncryptedPath: "/e4",       UnecryptedPath: "/f",     EncryptionKey: "key3"},
+}
+```
+
+Then if one asked for `sj://x/a/d`, the 2nd path would be the longest match and `"key2"` would be used to encrypt `d` and it would be appened to `"/e1"`. Similarly, if one asked for `sj://x/a/b/c/f`, then `"key2"` would be used to encrypt `b/c/f` and it would be appened to `"/e1"`. If one asked for `sj://x/f/g`, then `"key3"` would be used. `"key1"` would never be used as it is subsumed by the entry with `"key2"`.
+
+## Rationale
+
+This proposal is not a radical departure of the current method, just an extension to allowing a collection of paths to compose a bucket at any unencrypted spot. Thus, it shoudn't require any changes on the server side, reducing risk and allowing for earlier shipping.
+
+We could allow for "longest" matching. For example, in the above set of revealed paths one could use `"key1"` for `sj://x/a/b/c/d`. This begs a question about consistency. In principle be able to derive `"key1"` from the information provided by `"key2"`. If they disagree, an error could be thrown. This design takes the position that, these issues are trivial because it is very unlikely that such a share would be constructed, and even if it were, it would disagree.
```

Comment:
Also, I think "longest matching" share selection will not always give the right answer. Imagine I have macaroon A giving me read access to `a/b/`, and macaroon B giving me read/write access to `a/b/c/`, but macaroon B also has some other caveats. If B's other caveats are not satisfied, I should still be able to read the file `a/b/c/d` using macaroon A.

Reply:
So that points out a restriction the data model. There's only one macaroon per context (now "scope"). If you want access to the same project under different macaroons, that's a different scope.

Reply:
Ah. Right. I had the wrong model in my head.

But so this seems a little unwieldy, if people use it in the way I imagine they will. But nobody else seems to have the same expectations, so maybe it's fine.

But wouldn't it be nice if `uplink` could intelligently select which scope[s] _and_ which encryption key[s] to use, based on what you ask it to do? If we are already going to require an 'import' operation for each share, it seems like an obvious feature to provide. (Assuming, again, that we can add some kind of syntax for specifying a project in `sj://` URLs.)

Reply:
Project & bucket isn't unique enough because project names aren't globally unique across satellites, so you'd have to specify quite a bit of information for it to automatically select it in some cases.

I think this proposal is a middle ground in that it does the right thing for selecting buckets, but requires you to be explicit about what project/access level you're interacting with.

Reply:
Right, I get that, but project name is highly likely to be unique for all projects accessible to a given user, and if they're not, the user always has the option of referring to them by their UUID (or a prefix thereof).

Assuming some sort of inspection command that shows them the names and UUIDs of projects they have access to, along with details of the specific shares, that seems relatively easy to understand from a user perspective and easy for us to code.